File: bibtex.go
package main

import (
	"fmt"
	"io"
	"strings"
	"unicode"

	"golang.org/x/text/runes"
	"golang.org/x/text/transform"
	"golang.org/x/text/unicode/norm"
)

var asciiFolder = transform.Chain(norm.NFD, runes.Remove(runes.In(unicode.Mn)), norm.NFC)

// bibAsciiFold strips Unicode combining marks (accents, umlauts, etc.) so
// that "Müller" becomes "Muller" and "Émile" becomes "Emile". Used for
// citation key generation only; displayed author text is unaffected.
func bibAsciiFold(s string) string {
	out, _, _ := transform.String(asciiFolder, s)
	return out
}

// writeBibEntry writes one @article{...} block to w. used tracks emitted
// citation keys so collisions get _2, _3 suffixes within a single export.
func writeBibEntry(w io.Writer, p Paper, used map[string]int) {
	key := bibKey(p, used)
	fmt.Fprintf(w, "@article{%s,\n", key)
	if p.Authors != "" {
		fmt.Fprintf(w, "  author  = {%s},\n", bibAuthors(p.Authors))
	}
	if p.Title != "" {
		fmt.Fprintf(w, "  title   = {{%s}},\n", bibEscape(p.Title))
	}
	if p.Journal != "" {
		fmt.Fprintf(w, "  journal = {%s},\n", bibEscape(p.Journal))
	}
	if p.Year != "" {
		fmt.Fprintf(w, "  year    = {%s},\n", p.Year)
	}
	if p.Volume != "" {
		fmt.Fprintf(w, "  volume  = {%s},\n", bibEscape(p.Volume))
	}
	if p.Pages != "" {
		fmt.Fprintf(w, "  pages   = {%s},\n", bibEscape(p.Pages))
	}
	fmt.Fprintf(w, "  doi     = {%s}\n", p.DOI)
	fmt.Fprint(w, "}\n\n")
}

func bibKey(p Paper, used map[string]int) string {
	var first string
	if p.Authors != "" {
		if i := strings.IndexAny(p.Authors, " ,"); i > 0 {
			first = p.Authors[:i]
		} else {
			first = p.Authors
		}
	}
	var titleWord string
	for _, word := range strings.Fields(p.Title) {
		if alpha := bibAlpha(word); alpha != "" {
			titleWord = alpha
			break
		}
	}
	base := strings.ToLower(bibAlpha(first) + p.Year + titleWord)
	if base == "" {
		base = "paper"
	}
	used[base]++
	if used[base] == 1 {
		return base
	}
	return fmt.Sprintf("%s_%d", base, used[base])
}

// bibAlpha keeps only ASCII letters and digits, after folding Latin
// diacritics to their base letters.
func bibAlpha(s string) string {
	s = bibAsciiFold(s)
	var b strings.Builder
	for _, r := range s {
		if (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') || (r >= '0' && r <= '9') {
			b.WriteRune(r)
		}
	}
	return b.String()
}

// bibAuthors converts the stored "Family I., Family I." form to the
// "Family, I. and Family, I." form that BibTeX parses correctly. The split
// point is the first whitespace-separated token shaped like an initial
// (<rune>.) so multi-word surnames ("van der Berg") stay intact.
func bibAuthors(s string) string {
	parts := strings.Split(s, ", ")
	for i, p := range parts {
		tokens := strings.Fields(p)
		split := -1
		for j, t := range tokens {
			if isInitial(t) {
				split = j
				break
			}
		}
		if split > 0 {
			parts[i] = strings.Join(tokens[:split], " ") + ", " + strings.Join(tokens[split:], " ")
		}
	}
	return strings.Join(parts, " and ")
}

func isInitial(t string) bool {
	r := []rune(t)
	return len(r) == 2 && r[1] == '.'
}

func bibEscape(s string) string {
	r := strings.NewReplacer(
		`\`, `\textbackslash{}`,
		`&`, `\&`,
		`%`, `\%`,
		`_`, `\_`,
		`#`, `\#`,
		`$`, `\$`,
		`{`, `\{`,
		`}`, `\}`,
	)
	return r.Replace(s)
}

File: bibtex.py
import unicodedata
from typing import TextIO

from db import Paper


_ESCAPE_TABLE = {
    "\\": r"\textbackslash{}",
    "&": r"\&",
    "%": r"\%",
    "_": r"\_",
    "#": r"\#",
    "$": r"\$",
    "{": r"\{",
    "}": r"\}",
}


def write_bib_entry(out: TextIO, p: Paper, used: dict) -> None:
    """Write one @article{...} block. `used` tracks emitted citation keys
    so collisions within a single export get _2, _3 suffixes."""
    key = bib_key(p, used)
    out.write(f"@article{{{key},\n")
    if p.authors:
        out.write(f"  author  = {{{bib_authors(p.authors)}}},\n")
    if p.title:
        out.write(f"  title   = {{{{{bib_escape(p.title)}}}}},\n")
    if p.journal:
        out.write(f"  journal = {{{bib_escape(p.journal)}}},\n")
    if p.year:
        out.write(f"  year    = {{{p.year}}},\n")
    if p.volume:
        out.write(f"  volume  = {{{bib_escape(p.volume)}}},\n")
    if p.pages:
        out.write(f"  pages   = {{{bib_escape(p.pages)}}},\n")
    out.write(f"  doi     = {{{p.doi}}}\n}}\n\n")


def bib_key(p: Paper, used: dict) -> str:
    """firstAuthorSurname + year + firstTitleWord, lowercased, ASCII-folded.
    Collisions get _2, _3 suffixes via `used`."""
    first = p.authors
    if p.authors:
        candidates = [i for i in (p.authors.find(" "), p.authors.find(",")) if i > 0]
        if candidates:
            first = p.authors[:min(candidates)]

    title_word = ""
    for word in p.title.split():
        alpha = bib_alpha(word)
        if alpha:
            title_word = alpha
            break

    base = (bib_alpha(first) + p.year + title_word).lower()
    if not base:
        base = "paper"

    used[base] = used.get(base, 0) + 1
    return base if used[base] == 1 else f"{base}_{used[base]}"


def bib_alpha(s: str) -> str:
    """Keep ASCII letters/digits after ASCII-folding."""
    return "".join(c for c in bib_ascii_fold(s) if c.isascii() and c.isalnum())


def bib_ascii_fold(s: str) -> str:
    """NFD-normalize and strip combining marks: 'Müller' -> 'Muller'.

    Used for citation key generation only; displayed author text is
    unaffected.
    """
    return "".join(
        c for c in unicodedata.normalize("NFD", s)
        if unicodedata.category(c) != "Mn"
    )


def bib_authors(s: str) -> str:
    """Convert stored "Family I., Family I." to BibTeX
    "Family, I. and Family, I." The split point is the first
    whitespace-separated token shaped like an initial (<rune>.) so
    multi-word surnames ("van der Berg") stay intact.
    """
    out = []
    for part in s.split(", "):
        tokens = part.split()
        split = -1
        for j, t in enumerate(tokens):
            if _is_initial(t):
                split = j
                break
        if split > 0:
            out.append(" ".join(tokens[:split]) + ", " + " ".join(tokens[split:]))
        else:
            out.append(part)
    return " and ".join(out)


def _is_initial(t: str) -> bool:
    return len(t) == 2 and t[1] == "."


def bib_escape(s: str) -> str:
    """Escape \\ & % _ # $ { } for TeX. Single-pass so braces inside a
    `\\textbackslash{}` replacement aren't themselves re-escaped."""
    return "".join(_ESCAPE_TABLE.get(c, c) for c in s)
